### DISCUSSION TITLE
YAML configuration options documentation

### DISCUSSION THREAD

**Discussion by crides** [pyocd/pyOCD#1416] [Q&A]
What options are available in the YAML config? They aren't listed in the config docs on the website. I'm tryint to set things like RTT control block address and range and log level, but they aren't actually effective. Are these not implemented in the tool, or are they under other names (I'm using `address` `size` and `log-level` respectively)?

**Comment by flit** [comment]
Options are documented in the [session options reference](https://pyocd.io/docs/options.html).

Unfortunately, the RTT settings aren't available as session options; they can only be set via the command line. Adding these would be a great quick/first contribution from someone. 😄 (The RTT command was a quick implementation by a contributor. There is a lot of room for improvement!)

I've created #1417 to track the RTT options issue.

**Reply by crides** [reply]
Oh it's in the sessions options .. I kept looking for things with "config" in it :smile:. I'll look more into how the configs are fetched, and I may make a PR after that.

**Reply by flit** [reply]
The docs have just been updated so there's now a link to the session options reference from the configuration page.

Options are pretty easy to use. The [`Session`](https://github.com/pyocd/pyOCD/tree/main/pyocd/core/session.py) object has an `options` property that is a dict-like object ([`OptionsManager`](https://github.com/pyocd/pyOCD/tree/main/pyocd/core/options_manager.py)) used to access options. You don't need to do anything with config files; that's handled automatically. Add new options to the table in [`options.py`](https://github.com/pyocd/pyOCD/tree/main/pyocd/core/options.py) so they're documented and get a default.

Normally you pass the command line arg values to the helper when creating the session. If they are None, then they will be overridden by `-O` args, config file entries, and the option's default (in that order). Then just read the value through `session.options.get('foo')` (or `session.options['foo']`).

Int-typed options should have that type in the `options.py` entry, and be converted to int before being passed to the session.

**Reply by crides** [reply]
Okay, did a very simple patch in #1418. Also what are the fixed addresses mentioned in #1417?

**Reply by flit** [reply]
Oops, sorry I didn't see your reply! By fixed address I meant an option for the user to specify a known address for the RTT control block, to avoid a search. Not sure how realistic or useful that is… Another, probably better, possibility is using a symbol lookup from an ELF file.

**Reply by crides** [reply]
It does seem like the ELF contains the symbol for the control block. However, the JLink software is somehow able to obtain the location of the CB very fast, without the knowledge of an ELF file or configuring the search address range. I wonder what they use to achieve this

**Reply by flit** [reply]
If you mean the JLink gdbserver, then gdb gives it a chance to ask for symbol addresses just after connection (pyocd also uses this for RTOS thread awareness). But for the standalone JLink RTT viewer, I don't know! 🤷🏽 It would probably be possible to see what search pattern they use by enabling transfer logging before opening/connecting the RTT viewer.
